### PR TITLE
Feature/booking link

### PR DESCRIPTION
**Pull request by st3v3nhunt**
There are a few things that still need to be done. Some as part of this PR and others can be left to separate PRs - possibly?
- Casing of the ODS code is important and will break if it doesn't match the file (all uppercase). My thoughts are this should be done first thing on the request.
- On Choices, the website URL is used by systems marked as `EMIS (I)` and `INPS (I)`. At the moment all that happens is a message is displayed about not being able to book online. This should be replaced with the link to their website but that will need some refactoring on the opening times parser as the information is contained on the overview page.
- The loading of the POMI data probably wants moving somewhere better.
- The gp middleware feels like it is getting a bit big now - possibly time to move it into separate files?
- `getBookOnlineLink` is dependant upon `req.params.gpId` - i've not validated it in the function as I think this is best suited to being done when the request is first received along with the validation (or alteration) of the casing.

### CODE REVIEW

**Comment by st3v3nhunt** [comment]
The issue of casing has been 'fixed' in PR #78. That merges into `master`. PR #79 merges into the base branch for this PR. It might be easier to look at 79 over 78.

**Comment by st3v3nhunt** [comment]
For point 4, the validation of `req.params.gpId` there isn't much need as it is already handled. The route will not activate if there is no id and it results in the application level 404. If there is a pattern for ODS codes it might be worth validating it against that...

The pattern looks like it is available under the 'G Frame' section http://www.datadictionary.nhs.uk/data_dictionary/attributes/o/org/organisation_code_de.asp?shownav=1 and there are some default values http://www.datadictionary.nhs.uk/data_dictionary/data_field_notes/g/general_medical_practice_code_(patient_registration)_de.asp?shownav=1

**Comment by st3v3nhunt** [comment]
GP on Choices with different providers:
- Informatica - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=J82069
- EMIS - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=M85020
- Microtest - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=P88008
- NK - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=K81070
- TPP - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=M88628
- INPS - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=A82047
- EMIS (I) - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=M81037
- INPS (I) - http://www.nhs.uk/Services/GP/Overview/DefaultView.aspx?id=L85607

File: app/middleware/gp.js [hunk 1]
```diff
@@ -83,9 +84,40 @@ function getUrl(req, res, next) {
...
+function getBookOnlineLink(req, res, next) {
+  const gpId = req.params.gpId;
+  const systemSupplier = cache.get(gpId).supplier_name;
+
+  switch (systemSupplier) {
+    case 'EMIS':
+      // eslint-disable-next-line no-param-reassign
+      req.gpDetails.bookOnlineLink = 'https://patient.emisaccess.co.uk/Account/Login';
+      break;
+    case 'Informatica':
```
Comment (neilbmclaughlin) [inline app/middleware/gp.js:102]: What is the difference between, for example, the Informatica case and EMIS(I)?

Also, I think we need to discuss with Joe what he expects when a direct link to booking is not possible and what the rational was for the decisions made in Choices which made them put a link to things other than a booking page.

File: test/unit/middleware.js [hunk 2]
```diff
@@ -145,4 +146,91 @@ describe('Middleware', () => {
...
+      it('should handle casing', () => {
+      });
+    });
+    // TODO: There is another set of tests to be added here for when there is
+    // a website URL
+    describe('for known self suppliers with no website URL', () => {
+      const gpId = '123456';
+      const fakeRequest = {
+        params: { gpId },
+        gpDetails: { website: '' },
```
Comment (neilbmclaughlin) [inline test/unit/middleware.js:168]: Not sure why website is populated - it doesn't seem to be used.

File: app/middleware/gp.js [hunk 3]
```diff
@@ -83,9 +84,40 @@ function getUrl(req, res, next) {
...
 }
 
+function getBookOnlineLink(req, res, next) {
+  const gpId = req.params.gpId;
+  const systemSupplier = cache.get(gpId).supplier_name;
+
+  switch (systemSupplier) {
+    case 'EMIS':
+      // eslint-disable-next-line no-param-reassign
+      req.gpDetails.bookOnlineLink = 'https://patient.emisaccess.co.uk/Account/Login';
```
Comment (neilbmclaughlin) [inline app/middleware/gp.js:100]: Could we store these URL's in a CSV file and load into the cache too?

File: test/unit/server.js [hunk 4]
```diff
@@ -15,14 +16,16 @@ describe('Server', () => {
   const requestRoute = /\/organisations\/gppractices\/odscode\/[0-9]+.xml\?apikey=[a-z]*/;
   const requestRoute2 = /\/organisations\/gppractices\/[0-9]+\/overview.xml\?apikey=[a-z]*/;
```
Comment (neilbmclaughlin) [inline test/unit/server.js:17]: I named them but we should be able to come up with better names.

**Comment by st3v3nhunt** [comment]
Rebase this

**Pull request merged**